src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs

```diff
@@ -121,6 +180,29 @@ public DiagnosticAnalysisResult ToAggregatedForm()
             return new DiagnosticAnalysisResult(ProjectId, Version, DocumentIds, IsEmpty, FromBuild);
         }
 
+        public DiagnosticAnalysisResult UpdateAggregatedResult(VersionStamp version, DocumentId documentId, bool fromBuild)
+        {
+            return new DiagnosticAnalysisResult(ProjectId, version, DocumentIdsOrEmpty.Add(documentId), isEmpty: false, fromBuild: fromBuild);
+        }
+
+        public DiagnosticAnalysisResult Reset()
+        {
+            return new DiagnosticAnalysisResult(ProjectId, VersionStamp.Default, DocumentIds, IsEmpty, FromBuild);
+        }
+
+        public DiagnosticAnalysisResult DropExceptSyntax()
+        {
+            return new DiagnosticAnalysisResult(
+               ProjectId,
+               Version,
+               SyntaxLocals,
+               semanticLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+               nonLocals: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+               others: ImmutableArray<DiagnosticData>.Empty,
+               documentIds: null,
+               fromBuild: false);
```

Comment:
can we have comments about the 'fromBuild' being switched to 'false'?  it's unclear to me when that should be changing, and it's a subtlety i think a caller coudl get wrong.